Intermittent access violation

This is an issue to store the intermittent access violations that sometimes happens in the test runs. Primarily to see if it is always the same method etc.

Windows Python 3.10
```
Windows fatal exception: access violation

Current thread 0x00001108 (most recent call first):
  File "D:\a\apytypes\apytypes\lib\test\apyfloat\test_arithmetic.py", line 787 in test_long_power
```

Windows Python 3.12

```
Windows fatal exception: access violation

Current thread 0x0000051c (most recent call first):
  File "D:\a\apytypes\apytypes\lib\test\apyfloat\test_arithmetic.py", line 787 in test_long_power
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\python.py", line 195 in pytest_pyfunc_call
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_callers.py", line 102 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_manager.py", line 119 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_hooks.py", line 501 in __call__
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\python.py", line 1772 in runtest
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\runner.py", line 172 in pytest_runtest_call
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_callers.py", line 102 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_manager.py", line 119 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_hooks.py", line 501 in __call__
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\runner.py", line 240 in <lambda>
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\runner.py", line 340 in from_call
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\runner.py", line 239 in call_and_report
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\runner.py", line 134 in runtestprotocol
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\runner.py", line 115 in pytest_runtest_protocol
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_callers.py", line 102 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_manager.py", line 119 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_hooks.py", line 501 in __call__
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\main.py", line 364 in pytest_runtestloop
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_callers.py", line 102 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_manager.py", line 119 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_hooks.py", line 501 in __call__
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\main.py", line 339 in _main
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\main.py", line 285 in wrap_session
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\main.py", line 332 in pytest_cmdline_main
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_callers.py", line 102 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_manager.py", line 119 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\pluggy\_hooks.py", line 501 in __call__
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\config\__init__.py", line 174 in main
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\site-packages\_pytest\config\__init__.py", line 197 in console_main
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Scripts\pytest.exe\__main__.py", line 7 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main
.........................x
Error: Process completed with exit code 1.
```

--- a/lib/test/apyfixed/test_quantization.py
+++ b/lib/test/apyfixed/test_quantization.py
@@ -303,3 +303,25 @@ def test_issue_179():
     c = a.cast(bits=64, int_bits=64, quantization=QuantizationMode.RND)
     assert b.is_identical(APyFixed.from_float(0, bits=64, int_bits=64))
     assert c.is_identical(APyFixed.from_float(0, bits=64, int_bits=64))
+
+
+@pytest.mark.parametrize(
+    "mode",
+    [
+        QuantizationMode.TRN,
+        QuantizationMode.TRN_INF,
+        QuantizationMode.TRN_ZERO,
+        QuantizationMode.RND,
+        QuantizationMode.RND_INF,
+        QuantizationMode.RND_MIN_INF,
+        QuantizationMode.RND_ZERO,
+        QuantizationMode.RND_CONV,
+        QuantizationMode.RND_CONV_ODD,
+        QuantizationMode.JAM_UNBIASED,
+    ],
+)
+def test_quantize_away_all_bits(mode):
+    # These tests are used to trigger invalid memory accesses using Valgrind memcheck.
+    # Please leave them be.
+    a = APyFixed(0, bits=62002, int_bits=-12984)
+    assert float(a.cast(63, 3, quantization=mode)) == 0.0

--- a/src/apyfixed.cc
+++ b/src/apyfixed.cc
@@ -1164,9 +1164,11 @@ void APyFixed::_quantize_trn_inf(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (!limb_vector_is_negative(it_begin, it_end)
-            && limb_vector_or_reduce(it_begin, it_end, start_idx)) {
-            limb_vector_add_pow2(it_begin, it_end, start_idx);
+        if (start_idx < unsigned(_bits)) {
+            if (!limb_vector_is_negative(it_begin, it_end)
+                && limb_vector_or_reduce(it_begin, it_end, start_idx)) {
+                limb_vector_add_pow2(it_begin, it_end, start_idx);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
@@ -1186,9 +1188,11 @@ void APyFixed::_quantize_trn_zero(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (limb_vector_is_negative(it_begin, it_end)
-            && limb_vector_or_reduce(it_begin, it_end, start_idx)) {
-            limb_vector_add_pow2(it_begin, it_end, start_idx);
+        if (start_idx < unsigned(_bits)) {
+            if (limb_vector_is_negative(it_begin, it_end)
+                && limb_vector_or_reduce(it_begin, it_end, start_idx)) {
+                limb_vector_add_pow2(it_begin, it_end, start_idx);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
@@ -1208,8 +1212,10 @@ void APyFixed::_quantize_trn_mag(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (limb_vector_is_negative(it_begin, it_end)) {
-            limb_vector_add_pow2(it_begin, it_end, start_idx);
+        if (start_idx < unsigned(_bits)) {
+            if (limb_vector_is_negative(it_begin, it_end)) {
+                limb_vector_add_pow2(it_begin, it_end, start_idx);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
@@ -1229,7 +1235,9 @@ void APyFixed::_quantize_rnd(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+        if (start_idx <= unsigned(_bits)) {
+            limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+        }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
 }
@@ -1248,9 +1256,11 @@ void APyFixed::_quantize_rnd_zero(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (limb_vector_is_negative(it_begin, it_end)
-            || limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
-            limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+        if (start_idx <= unsigned(_bits)) {
+            if (limb_vector_is_negative(it_begin, it_end)
+                || limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
+                limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
@@ -1270,9 +1280,11 @@ void APyFixed::_quantize_rnd_inf(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (!limb_vector_is_negative(it_begin, it_end)
-            || limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
-            limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+        if (start_idx <= unsigned(_bits)) {
+            if (!limb_vector_is_negative(it_begin, it_end)
+                || limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
+                limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
@@ -1292,8 +1304,10 @@ void APyFixed::_quantize_rnd_min_inf(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
-            limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+        if (start_idx <= unsigned(_bits)) {
+            if (limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
+                limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
@@ -1313,9 +1327,11 @@ void APyFixed::_quantize_rnd_conv(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (limb_vector_test_bit(it_begin, it_end, start_idx)
-            || limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
-            limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+        if (start_idx < unsigned(_bits)) {
+            if (limb_vector_test_bit(it_begin, it_end, start_idx)
+                || limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
+                limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
@@ -1335,9 +1351,11 @@ void APyFixed::_quantize_rnd_conv_odd(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (!limb_vector_test_bit(it_begin, it_end, start_idx)
-            || limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
-            limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+        if (start_idx < unsigned(_bits)) {
+            if (!limb_vector_test_bit(it_begin, it_end, start_idx)
+                || limb_vector_or_reduce(it_begin, it_end, start_idx - 1)) {
+                limb_vector_add_pow2(it_begin, it_end, start_idx - 1);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }
@@ -1375,8 +1393,10 @@ void APyFixed::_quantize_jam_unbiased(
         limb_vector_lsl(it_begin, it_end, left_shift_amnt);
     } else {
         unsigned start_idx = -left_shift_amnt;
-        if (limb_vector_or_reduce(it_begin, it_end, start_idx)) {
-            limb_vector_set_bit(it_begin, it_end, start_idx, true);
+        if (start_idx < unsigned(_bits)) {
+            if (limb_vector_or_reduce(it_begin, it_end, start_idx)) {
+                limb_vector_set_bit(it_begin, it_end, start_idx, true);
+            }
         }
         limb_vector_asr(it_begin, it_end, start_idx);
     }

--- a/src/apytypes_util.h
+++ b/src/apytypes_util.h
@@ -783,6 +783,7 @@ template <class RANDOM_ACCESS_ITERATOR>
 }
 
 //! Reduce the first `n` bits in a limb vector over bitwise `or`. Returns bool.
+//! Undefined behaviour if bit `n` is located outside the limb vector
 template <class RANDOM_ACCESS_ITERATOR>
 [[maybe_unused, nodiscard]] static APY_INLINE bool limb_vector_or_reduce(
     RANDOM_ACCESS_ITERATOR cbegin_it, RANDOM_ACCESS_ITERATOR cend_it, unsigned n
@@ -811,7 +812,8 @@ template <class RANDOM_ACCESS_ITERATOR>
     return false;
 }
 
-//! Test if the `n` -th bit (zero indexed) is set in a limb vector
+//! Test if the `n` -th bit (zero indexed) is set in a limb vector. Undefined behaviour
+//! if bit `n` is located outside the limb vector.
 template <class RANDOM_ACCESS_ITERATOR>
 [[maybe_unused, nodiscard]] static APY_INLINE bool limb_vector_test_bit(
     RANDOM_ACCESS_ITERATOR cbegin_it, RANDOM_ACCESS_ITERATOR cend_it, unsigned n
@@ -825,7 +827,8 @@ template <class RANDOM_ACCESS_ITERATOR>
     return mask & limb;
 }
 
-//! Set the `n` -th bit (zero indexed) of a limb vector to `bit`
+//! Set the `n` -th bit (zero indexed) of a limb vector to `bit`. Undefined behaviour if
+//! bit `n` is located outside the limb vector.
 template <class RANDOM_ACCESS_ITERATOR>
 [[maybe_unused]] static APY_INLINE void limb_vector_set_bit(
     RANDOM_ACCESS_ITERATOR begin_it, RANDOM_ACCESS_ITERATOR end_it, unsigned n, bool bit

